### PR TITLE
Fully deserialize as necessary to look up pointers

### LINKED ISSUE #66: formatDescriptor() throws on certain deserialized descriptors with deeply-nested self-references
When `formatDescriptor()` is called on a deserialized descriptor of an object with a self-referential property at a nesting depth beyond `maxDepth`, it may throw a `PointerLookupError`.

MRE:
```javascript
const value = {};
const actualDescriptor = concordance.describe({
  a: {
    b: value
  },
  c: value
});

const serializedActual = concordance.serialize(actualDescriptor);
const deserializedActual = concordance.deserialize(serializedActual);

concordance.formatDescriptor(deserializedActual, {maxDepth: 1});
```
This throws:
```
  PointerLookupError {
    index: 3,
    message: 'Could not deserialize buffer: pointer 3 could not be resolved',
  }

  › mapPointerDescriptor (node_modules/concordance/lib/serialize.js:321:67)
  › node_modules/concordance/lib/serialize.js:336:14
  › deserializeRecord (node_modules/concordance/lib/serialize.js:255:12)
  › recursor (node_modules/concordance/lib/serialize.js:264:12)
  › deserializeComplex (node_modules/concordance/lib/metaDescriptors/property.js:18:17)
  › node_modules/concordance/lib/serialize.js:336:35
  › deserializeRecord (node_modules/concordance/lib/serialize.js:267:10)
  › Object.recursor [as next] (node_modules/concordance/lib/serialize.js:264:12)
  › Object.next [as recursor] (node_modules/concordance/lib/recursorUtils.js:59:44)
  › Object.formatDescriptor (node_modules/concordance/lib/format.js:74:35)
```

To reproduce the issue, the following must hold:
- The provided `maxDepth` is non-zero
- The deeply-nested self-reference occurs at a depth greater than `maxDepth`
- The shallow self-reference occurs at a depth at most `maxDepth`
- The name of the property containing the deeply-nested self-reference is lexically less than the name of the property containing the shallow self-reference

I believe the problem to be that the descriptor is deserialized lazily, and the formatter never calls upon the deeply-nested self-referential value, so when the shallow self-reference is called upon, its pointer references a descriptor that hasn't been deserialized yet.

I first encountered this in AVA, where it can cause an error to be thrown when a snapshot assertion comparing against such a value fails.

I'm working on a PR that may fix this presently.

### PR DESCRIPTION
**Pull request by ninevra**
Tests and hopefully fixes #66.

This PR attempts to address #66 by consuming to completion the recursor of any `DeserializedMixin` that would be elided due to an `ObjectFormatter` hitting `maxDepth`. This causes the entire value to be deserialized, ensuring that if the parent descriptor contains any pointers into the value, their targets will be valid.

This is my first look at this codebase. This might not be the right aproach at all. I might be missing other cases besides `formatDescriptor` and `diffDescriptors` that need tests and/or fixes. Hopefully I've at least got the basics like "what is a recursor" right.

There is definitely a better place to put these test cases, but I'm not sure where it is. The behavior under test is sort of an intersection of format, diff, serialize, and maxDepth, each of which has its own test suite.

### CODE REVIEW

**Comment by novemberborn** [comment]
Wow, thanks for the PR! It may take be some time to wrap my head around the changes. I'm impressed you got this far.

**Comment by novemberborn** [comment]
I think there's a different approach that's more clear, have a look at my commit.

Recursing seems to be lazy (I say "seems" because I wrote this code years ago). That's fine for comparisons, but not when formatting. And diffing is just more complicated formatting.

So instead, for snapshots, we should deserialize greedily. Not when we hit max depth but before we even format (or diff). This PR now adds that as an option, and then we should enable it in AVA here: https://github.com/avajs/ava/blob/378f605f26c6d290918a3817d950b6a6370f2a18/lib/snapshot-manager.js#L344

What do you think?

Note that I rewrote the recursing logic to use stacks, not recursion, since the other code also tries real hard to avoid recursion.

**Comment by ninevra** [comment]
Having an option to expose this behavior definitely makes it clearer what's happening and why.

I don't think the user should have to opt-in to greedy deserialization just to get correct formatting/diffing behavior, though; formatting really shouldn't ever throw when given a valid descriptor as input. Particularly since nothing about `concordance`'s public interface makes it clear that there's any meaningful difference between deserialized descriptors and those produced by `describe()`.

I'm also not sure users would ever benefit from knowing whether their descriptors are greedy or lazy.

Maybe greedy deserialization could be the default, with or without an "opt-out" `lazy` option? Or maybe deserialized descriptors could expose an option to _become_ greedy, which `formatDescriptor` and `diffDescriptors` would toggle on?

In AVA's case, [lib/snapshot-manager.js#L344](https://github.com/avajs/ava/blob/378f605f26c6d290918a3817d950b6a6370f2a18/lib/snapshot-manager.js#L344) is where the descriptor's deserialized, but it doesn't get formatted unless the assertion fails, and the formatting happens at [lib/assert.js#L13](https://github.com/avajs/ava/blob/378f605f26c6d290918a3817d950b6a6370f2a18/lib/assert.js#L13). If there's any benefit to lazy deserialization, one would want to select greedy deserialization only at the point where it's necessary, when printing the diff; but that part of the code doesn't know whether it's working with a deserialized descriptor or not. (And only `concordance` knows whether the descriptor is of a type that has recursive structure or pointers.)

Another codebase running into this problem might mix deserialized and "fresh" descriptors much more extensively, and then when running into this problem, would have to find every place where they deserialize and add `{greedy: true}`, even if most of those descriptors never actually got formatted.

File: lib/recursorUtils.js [hunk 1]
```diff
@@ -24,6 +24,22 @@ function fork (recursor) {
 }
 exports.fork = fork
 
+function greedy (recursor) {
```
Comment (ninevra) [inline lib/recursorUtils.js:27]: I'm definitely nitpicking/bikeshedding here, but I don't think `greedy` is the right name for this. `greedy(recursor)` sounds like it ought to return a recursor that's greedy, when it actually _greedily consumes_ the recursor and returns nothing. Using a verb name would better convey that to a reader.

Comment (novemberborn) [reply lib/recursorUtils.js:27]: `consumeDeep`?

**Comment by novemberborn** [comment]
Agreed, perhaps greedy should be the default, and `lazy` the alternative.

In AVA's case, for the snapshot to match, everything has to have been deserialized. And then if the snapshot doesn't match, we want full deserialization for formatting. So there is no performance benefit in deserializing lazily.

**Comment by ninevra** [comment]
Another option (besides defaulting to greedy deserialization) is to do greedy deserialization in a clause of the error-handling part of `mapPointerDescriptor`:
```javascript
if (!descriptorsByPointerIndex.has(descriptor.index)) {
  // Consume the whole descriptor to map any missed pointees
  if (typeof rootDescriptor.createRecursor === 'function') {
    recursorUtils.consumeDeep(rootDescriptor.createRecursor())
  }
  if (!descriptorsByPointerIndex.has(descriptor.index)) {
    throw new PointerLookupError(descriptor.index)
  }
}
```

**Comment by novemberborn** [comment]
@ninevra yes! That's brilliant. See the commit I just pushed.

**Comment by ninevra** [comment]
Looks good to me.

The codecov failure is probably negligible, it's just flagging the `throw new PointerLookupError()` line.

**Comment by novemberborn** [comment]
Released as 5.0.2 🥳

**Pull request merged**